src/estimagic/benchmarking/benchmark_reports.py

```diff
@@ -0,0 +1,171 @@
+    x_precision=1e-4,
+    y_precision=1e-4,
+):
+    """Create a DataFrame with rank information for a set of problems.
+
+    Args:
+        problems (dict): estimagic benchmarking problems dictionary. Keys are the
+            problem names. Values contain information on the problem, including the
+            solution value.
+        results (dict): estimagic benchmarking results dictionary. Keys are
+            tuples of the form (problem, algorithm), values are dictionaries of the
+            collected information on the benchmark run, including 'criterion_history'
+            and 'time_history'.
+        runtime_measure (str): "n_evaluations", "n_batches" or "walltime".
+            This is the runtime until the desired convergence was reached by an
+            algorithm. This is called performance measure by Moré and Wild (2009).
+            Default is "n_evaluations".
+        normalize_runtime (bool): If True the runtime each algorithm needed for each
+            problem is scaled by the time the fastest algorithm needed. If True, the
+            resulting plot is what Moré and Wild (2009) called data profiles.
+            Default is False.
+        stopping_criterion (str): one of "x_and_y", "x_or_y", "x", "y". Determines
+            how convergence is determined from the two precisions.
+        x_precision (float or None): how close an algorithm must have gotten to the
+            true parameter values (as percent of the Euclidean distance between start
+            and solution parameters) before the criterion for clipping and convergence
+            is fulfilled. Default is 1e-4.
+        y_precision (float or None): how close an algorithm must have gotten to the
+            true criterion values (as percent of the distance between start
+            and solution criterion value) before the criterion for clipping and
+            convergence is fulfilled. Default is 1e-4.
+
+    Returns:
+        pandas.DataFrame: columns are the algorithms, indexes are the problems.
+            The values are the ranks of the algorithms for each problem,
+            0 means algorithms was the fastest, 1 means it was the second fastest
+            and so on. If an algorithm did not converge on a problem, the value is
+            "failed". If an algorithm did encounter an error during optimization,
+            the value is "error".
+
+    """
+    histories, converged_info = process_benchmark_results(
+        problems=problems,
+        results=results,
+        stopping_criterion=stopping_criterion,
+        x_precision=x_precision,
+        y_precision=y_precision,
+    )
+    scenarios = list({algo[1] for algo in results.keys()})
+
+    success_info = converged_info.replace({True: "success", False: "failed"})
+    for key, value in results.items():
+        if isinstance(value["solution"], str):
+            success_info.at[key] = "error"
+
+    solution_times = histories.groupby(["problem", "algorithm"])[runtime_measure].max()
```

Comment:
This is duplicated with code in `profile_plot._create_solution_times`, but I think it forgets to set the solution time to infinity for unsolved problems which would potentially lead to very wrong results. We should probably make `_create_solution_times` public and use it here.

Reply:
I can do that. From a DRY perspective, I understand the virtue of using ```profile_plot._create_solution_times``` here. Two notes, however, why I decided not to call it directly:

1) We need the ```solution_times``` DataFrame to be in a different format. Calling the ```profile_plot._create_solution_times``` function would require to apply ```solution_times.stack().reset_index()``` 
to the resulting DataFrame.

2) What ```profile_plot._create_solution_times``` does with ```solution_times[~converged_info] = np.inf```, is done [here](https://github.com/OpenSourceEconomics/estimagic/blob/benchmark_reports/src/estimagic/benchmarking/benchmark_reports.py#L139). ```success_info``` contains one of the strings "success", "failed", "error" as values for each benchmark problem. Hence, problems that did not converge are not assigned an integer rank, but the string "failed" or "error". For comparison, in  ```profile_plot._create_solution_times``` ```np.inf``` refers to either "failed" or "error".